src/riscv.c

```diff
@@ -224,76 +298,105 @@ riscv_t *rv_create(riscv_user_t rv_attr)
-
-        const struct Elf32_Sym *end;
-        if ((end = elf_get_symbol(elf, "_end")))
-            attr->break_addr = end->st_value;
-
-        assert(elf_load(elf, attr->mem));
-
-        /* set the entry pc */
-        const struct Elf32_Ehdr *hdr = get_elf_header(elf);
-        assert(rv_set_pc(rv, hdr->e_entry));
-
-        elf_delete(elf);
-
-        /* this variable has external linkage to mmu_io defined in system.c */
-        extern riscv_io_t mmu_io;
-        memcpy(&rv->io, &mmu_io, sizeof(riscv_io_t));
+    /* this variable has external linkage to mmu_io defined in system.c */
+    extern riscv_io_t mmu_io;
+    /* install the MMU I/O handlers */
+    memcpy(&rv->io, &mmu_io, sizeof(riscv_io_t));
+#else
+    /* install the I/O handlers */
+    const riscv_io_t io = {
+        /* memory read interface */
+        .mem_ifetch = MEMIO(ifetch),
+        .mem_read_w = MEMIO(read_w),
+        .mem_read_s = MEMIO(read_s),
+        .mem_read_b = MEMIO(read_b),
+
+        /* memory write interface */
+        .mem_write_w = MEMIO(write_w),
+        .mem_write_s = MEMIO(write_s),
+        .mem_write_b = MEMIO(write_b),
+
+        /* system services or essential routines */
+        .on_ecall = ecall_handler,
+        .on_ebreak = ebreak_handler,
+        .on_memcpy = memcpy_handler,
+        .on_memset = memset_handler,
+        .on_trap = trap_handler,
+    };
+    memcpy(&rv->io, &io, sizeof(riscv_io_t));
+#endif /* RV32_HAS(SYSTEM) */
+
+#else
+    /* *-----------------------------------------*
+     * |              Memory layout              |
+     * *----------------*----------------*-------*
+     * |  kernel image  |  initrd image  |  dtb  |
+     * *----------------*----------------*-------*
+     */
+
+    char *ram_loc = (char *) attr->mem->mem_base;
+    map_file(&ram_loc, attr->data.system.kernel);
+
+    uint32_t dtb_addr = attr->mem->mem_size - (1 * 1024 * 1024);
+    ram_loc = ((char *) attr->mem->mem_base) + dtb_addr;
+    map_file(&ram_loc, attr->data.system.dtb);
+    /*
+     * Load optional initrd image at last 8 MiB before the dtb region to
+     * prevent kernel from overwritting it
+     */
+    if (attr->data.system.initrd) {
+        uint32_t initrd_addr = dtb_addr - (8 * 1024 * 1024);
+        ram_loc = ((char *) attr->mem->mem_base) + initrd_addr;
+        map_file(&ram_loc, attr->data.system.initrd);
     }
-#endif /* SYSTEM */
+
+    /* this variable has external linkage to mmu_io defined in system.c */
+    extern riscv_io_t mmu_io;
+    memcpy(&rv->io, &mmu_io, sizeof(riscv_io_t));
+
+    /* setup RISC-V hart */
+    rv_set_reg(rv, rv_reg_a0, 0);
+    rv_set_reg(rv, rv_reg_a1, dtb_addr);
+
+    /* setup timer */
+    attr->timer = 0xFFFFFFFFFFFFFFF;
+
+    /* setup PLIC */
+    attr->plic = plic_new();
+    assert(attr->plic);
+    attr->plic->rv = rv;
+
+    /* setup UART */
+    attr->uart = u8250_new();
+    assert(attr->uart);
+    attr->uart->in_fd = 0;
+    attr->uart->out_fd = 1;
```

Comment:
Replace to 
```cpp
    attr->uart->in_fd = STDIN_FILENO;
    attr->uart->out_fd = STDOUT_FILENO;
```

Reply:
I think we can leverage the `fd_stdin` and `fd_stdout` for better compatibility.

Reply:
Ah yes. This will be better.